fix friendly url data providers

vitek-rostislav opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| mismatch in variables assignments reported in #945 
|New feature| No <!-- Do not forget to update docs/ -->
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| closes #945, closes #1174 <!-- Write "closes #123" for the issue to be closed automatically during merge -->
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

grossmannmartin commented:
Hi, nice. 
I'm curious, why this PR shows 60 commits it's actually just one, could rebase help?
Also, in the issue #945 is a comment from Jan, could you please answer him? I believe there is no problem.

But approve :+1:

vitek-rostislav commented:
Hi, thanks for the review :+1: 

> I'm curious, why this PR shows 60 commits it's actually just one, could rebase help?

Yes, it was my mistake (the PR was targeted against master originally). I rebased the branch on `8.0` :wink: 

> Just wondering about upgrade notes. Should there be any?

:thinking: Hm, I have removed one method from `FriendlyUrlDataFactory` - there should be an upgrade note. I have added it in a fixup commit. For the other changes, I do not think any upgrade notes are needed

grossmannmartin commented:
### Tests
I did some testing and generating of URL for brands and articles is okay, but I found another error in
`ProductListFriendlyUrlDataProvider` and `ProductDetailFriendlyUrlDataProvider` classes.

in both of classes near line 50 is following ` ->setParameter('locale', $domainConfig->getId())`, resulting in empty data from query because there is no locale named 1 (for instance :))
It should be like ` ->setParameter('locale', $domainConfig->getLocale())`

Currently, product detail and product list friendly URLs are not generated.

PetrHeinz commented:
@vitek-rostislav I'm wondering why you haven't fixed the actual implementation of the `FriendlyUrlDataFactory::createFromData()` method and use it instead of fixing the implementation in the four places (which can be seen as a [DRY violation](https://github.com/shopsys/shopsys/blob/v7.2.0/docs/contributing/code-quality-principles.md#dont-repeat-yourself))...

As a side-effect, this would also remove the need for the BC break.

vitek-rostislav commented:
Hi @PetrHeinz, thanks for pointing this out, for some reason, this has not occurred to me :man_facepalming: I will fix it this way :+1:

grossmannmartin commented on packages/framework/src/Model/Article/ArticleDetailFriendlyUrlDataProvider.php:
```diff
@@ -54,9 +54,20 @@ public function getFriendlyUrlData(DomainConfig $domainConfig): array
-            $friendlyUrlData = $this->friendlyUrlDataFactory->create();
-            $friendlyUrlData->name = $data['id'];
-            $friendlyUrlData->id = $data['name'];
+            if (method_exists($this->friendlyUrlDataFactory, 'createFromIdAndName')) {
```
`method_exists` in a loop seems inefficient to me

grossmannmartin commented on packages/framework/src/Model/Article/ArticleDetailFriendlyUrlDataProvider.php:
```diff
@@ -54,9 +54,20 @@ public function getFriendlyUrlData(DomainConfig $domainConfig): array
+                $friendlyUrlData = $this->friendlyUrlDataFactory->create();
+                $friendlyUrlData->id = $data['id'];
+                $friendlyUrlData->name = $data['name'];
+            }
```
The whole method is quite hard to read. Can you please extract some parts into methods?
I thought about 
```php 
if (method_exists($this->friendlyUrlDataFactory, 'createFromIdAndName')) {
    return $this->getDataTheNewWay();
} else {
    return $this->getDataTheOldFashionWay();
}
```
I know there will be for loop in both methods, but It helps extensibility and readability of a code.
What do you think?

_Same applies for all the Data providers_

vitek-rostislav replied:
fair point, fixed

grossmannmartin commented on packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlDataFactoryInterface.php:
```diff
@@ -2,6 +2,9 @@
 namespace Shopsys\FrameworkBundle\Component\Router\FriendlyUrl;
 
+/**
+ * @method FriendlyUrlData createFromIdAndName(int $id, string $name)
```
It's to maintain backward compatibility? Can you please mark it somehow, so when we remove deprecations we don't forget to update the interface as this should definitely be part of the interface
(and I don't really know how properly)

vitek-rostislav replied:
yes, for BC. I believe the @method annotation is sufficient enough - I have added a note to the user story that solves deprecations for 8.0

grossmannmartin commented on packages/framework/src/Model/Article/ArticleDetailFriendlyUrlDataProvider.php:
```diff
@@ -54,9 +54,20 @@ public function getFriendlyUrlData(DomainConfig $domainConfig): array
+                );
+                $friendlyUrlData = $this->friendlyUrlDataFactory->create();
+                $friendlyUrlData->id = $data['id'];
+                $friendlyUrlData->name = $data['name'];
```
Why don't you use `createFromData` method? You can omit the deprecations here and have it only in `FriendlyUrlDataFactory`

vitek-rostislav replied:
I was not able to use it as it is not part of the interface...But your point makes sense, rewritten by adding the @method annotation along with @deprecated sign into the interface :grin:   :+1:

grossmannmartin commented on packages/framework/src/Model/Article/ArticleDetailFriendlyUrlDataProvider.php:
```diff
@@ -70,4 +74,22 @@ public function getRouteName(): string
+     * @param array $data
+     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
+     */
+    protected function createFriendlyUrlDataNewWay(array $data): FriendlyUrlData
```
what about `createFriendlyUrlDataNewWay(int $id, string $name): FriendlyUrlData` ?

vitek-rostislav replied:
we discussed it - it is not important as the method will be removed in `8.0` (added @internal annotation to ensure that)

grossmannmartin commented on packages/framework/src/Model/Product/Brand/BrandDetailFriendlyUrlDataProvider.php:
```diff
@@ -55,10 +56,13 @@ public function getFriendlyUrlData(DomainConfig $domainConfig): array
-            $friendlyUrlData->name = $data['id'];
-            $friendlyUrlData->id = $data['name'];
+            if ($newMethodForDataCreationExists) {
+                $friendlyUrlData = $this->createFriendlyUrlDataNewWay($data);
```
I'm not really sure about the names... Unfortunately I don't have better names...

vitek-rostislav replied:
we discussed it - it is not important as the methods will be removed in `8.0` (added @internal annotation to ensure that)

PetrHeinz commented on packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlDataFactoryInterface.php:
```diff
@@ -2,6 +2,9 @@
 namespace Shopsys\FrameworkBundle\Component\Router\FriendlyUrl;
 
+/**
+ * @method FriendlyUrlData createFromIdAndName(int $id, string $name)
```
Please add a note about this into *US-6091 Resolving of deprecations for v8.0.0* in our internal backlog, I'm bit worried the method would not be added into the interface otherwise.

vitek-rostislav replied:
I have already added a note there that it is necessary to resolve `@method` and `@internal` annotations, is it sufficient?

PetrHeinz replied:
Well this is the only `@method` that needs any kind of resolving, but fine

PetrHeinz commented on packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactoryInterface.php:
```diff
@@ -2,7 +2,7 @@
 namespace Shopsys\FrameworkBundle\Component\Router\FriendlyUrl;
 
-interface FriendlyUrlFactoryInterface
+interface   FriendlyUrlFactoryInterface
```
```suggestion
interface FriendlyUrlFactoryInterface
```
